collector/bgp.go

```diff
@@ -8,155 +8,88 @@ import (
 
-	bgpErrors           = []error{}
-	totalBGPErrors      = 0.0
-	bgp6Errors          = []error{}
-	totalBGP6Errors     = 0.0
-	bgpL2VPNErrors      = []error{}
-	totalBGPL2VPNErrors = 0.0
-
 	bgpPeerTypes          = kingpin.Flag("collector.bgp.peer-types", "Enable the frr_bgp_peer_types_up metric (default: disabled).").Default("False").Bool()
 	frrBGPDescKey         = kingpin.Flag("collector.bgp.peer-types.keys", "Select the keys from the JSON formatted BGP peer description of which the values will be used with the frr_bgp_peer_types_up metric. Supports multiple values (default: type).").Default("type").Strings()
 	bgpPeerDescs          = kingpin.Flag("collector.bgp.peer-descriptions", "Add the value of the desc key from the JSON formatted BGP peer description as a label to peer metrics. (default: disabled).").Default("False").Bool()
 	bgpPeerDescsText      = kingpin.Flag("collector.bgp.peer-descriptions.plain-text", "Use the full text field of the BGP peer description instead of the value of the JSON formatted desc key (default: disabled).").Default("False").Bool()
 	bgpAdvertisedPrefixes = kingpin.Flag("collector.bgp.advertised-prefixes", "Enables the frr_exporter_bgp_prefixes_advertised_count_total metric which exports the number of advertised prefixes to a BGP peer. This is an option for older versions of FRR that don't have PfxSent field (default: disabled).").Default("False").Bool()
 )
 
-// BGPCollector collects BGP metrics, implemented as per prometheus.Collector interface.
-type BGPCollector struct{}
-
-// NewBGPCollector returns a BGPCollector struct.
-func NewBGPCollector() *BGPCollector {
-	return &BGPCollector{}
+func init() {
+	registerCollector(bgpSubsystem, enabledByDefault, NewBGPCollector)
+	registerCollector(bgpSubsystem+"6", enabledByDefault, NewBGP6Collector)
+	registerCollector(bgpSubsystem+"l2vpn", enabledByDefault, NewBGPL2VPNCollector)
 }
 
-// Name of the collector. Used to populate flag name.
-func (*BGPCollector) Name() string {
-	return bgpSubsystem
-}
-
-// Help describes the metrics this collector scrapes. Used to populate flag help.
-func (*BGPCollector) Help() string {
-	return "Collect BGP Metrics"
-}
-
-// EnabledByDefault describes whether this collector is enabled by default. Used to populate flag default.
-func (*BGPCollector) EnabledByDefault() bool {
-	return true
-}
-
-// Describe implemented as per the prometheus.Collector interface.
-func (*BGPCollector) Describe(ch chan<- *prometheus.Desc) {
-	for _, desc := range getBgpDesc() {
-		ch <- desc
-	}
+type bgpCollector struct {
+	logger log.Logger
 }
 
-// Collect implemented as per the prometheus.Collector interface.
-func (c *BGPCollector) Collect(ch chan<- prometheus.Metric) {
-	bgpErrors = []error{}
-	collectBGP(ch, "ipv4")
+// NewBGPCollector collects BGP metrics, implemented as per the Collector interface.
+func NewBGPCollector(logger log.Logger) (Collector, error) {
+	return &bgpCollector{logger: logger}, nil
```

Comment:
(2) The whole code from `setBgpDesc` should be moved here, since it's enough to execute it once. Then `bgpDescMtx` is not needed any more (since you only create the data here and later only read it).

Reply:
Moving all code from `setBgpDesc` breaks the tests. I can't think of a cleaner way than leaving `setBgpDesc` in place. Thoughts?

`bgpDescMtx` has now been removed, though, thanks to (1).

Reply:
Your implementation to have `getBGPDesc` and call it in `NewBGPCollector` and `NewBGP6Collector` looks good to me.